docs/adr/0008-allowing-more-types-as-errors.md

```diff
@@ -0,0 +1,139 @@
+* The type can't be used exactly as it could in non-error contexts.
+
+ie, it's a bit of "either or, but not both". But it makes for more natural foreign code:
+
+```
+try:
+    do_it()
+except ErrorDetailsException as e:
+    # There's no `ErrorDetails` type in this world.
+    # ie, `e' is an  `ErrorDetailsException`
+    # `ErrorDetailsException` derives from `Exception`
+    e.message()
+```
+
+### B) Create an original type and an error type
+
+We create 2 types - the `interface` itself unchanged from how the struct would be generated now,
+and another exception "wrapper" for the type.
+It means you *catch* a slightly different name from the interface, but the non-error version of the interface is what you deal with.
+
+```
+try:
+    do_it()
+except ErrorDetailsException as e:
+    # There are 2 types here:
+    # `e` is an `ErrorDetailsException`
+    # `e.inner` (or any other name) is an `ErrorDetails` - does not derive from `Exception`
+    # We can make it look natural via delegation.
+    e.message()
+    # But also use the original interface
+    HandleErrorReportingProblem(e.inner)
+```
+
+To clarify, you could imagine generated Python code looking something like:
+```
+class ErrorDetails:
+    ...
+class ErrorDetailsException(Exception):
+    ...
+```
+### C) Create and original type with an error type nested within the original
+
+Very similar to the above - there are still 2 unique types created, but only one type is a top-level
+type - the error type is created inside the exception.
+
+The code catching exceptions might look something like:
+```
+try:
+    do_it()
+except ErrorDetails.Exception as e:
+    # same block above as for B
+```
+
+ie, there's no top-level `ErrorDetailsException` type, it's accessed via `ErrorDetails.Exception`
+
+To clarify, you could imagine generated Python code looking something like:
+```
+class ErrorDetails:
+    # snip regular interface.
+    class Exception(Exception):
+        # snip error specific code
+```
+
+## Decision Outcome
+
+Chosen option: "[option X]", because [justification. e.g., only option, which meets k.o. criterion decision driver | which resolves force force | … | comes out best (see below)].
+
+## Pros and Cons of the Options
+
+### 1: Do nothing - don't support interfaces as errors.
+
+Con: Seems a bit negative :)
+
+### 2: Implement Option A above
+
+- Pro: no magic, there's exactly 1 type for the type in the UDL/macro
+- Con: That type gets "tainted" by being used as an error. Using the interface in other contexts is less natural.
+- Con: It might be tricky to use as an external type - all external consumers must use as an error, or none of them must.
+
+### 3: Implement Option B above:
+
+- Pro: a more "pure" system - the interface can be used naturally in contexts other than being an error.
+- Pro: It mirrors Rust features in that it "adds" a capability rather then transforms it.
```

Comment:
I agree that this feels more pure/natural from a Rust POV.  But in practice, aren't error types treated somewhat specially by the library?

Reply:
I added a few comments around this.